### ISSUE TITLE
IfcRelSpaceBoundary : geom.create_shape fails

### ISSUE THREAD

**Issue by CyrilWaechter** [IfcOpenShell/IfcOpenShell#635]
Hi,
I tried to use `geom.create_shape` on an IfcRelSpaceBoundary which has no `Representation`. I tried multiple things but failed to create a geometry following your clues in #165  : 
```python
# coding: utf8
import ifcopenshell
import ifcopenshell.geom

ifc_path = "TestModel.ifc"
ifc_file = ifcopenshell.open(ifc_path)
space_boundaries = ifc_file.by_type('IfcRelSpaceBoundary')
s = geom.settings()
# settings.set(settings.USE_PYTHON_OPENCASCADE, True)
s.set(s.USE_WORLD_COORDS, True)
s.set(s.EXCLUDE_SOLIDS_AND_SURFACES, False)
s.set(s.INCLUDE_CURVES, True)

# occ_display = geom.utils.initialize_display()
for boundary in (b for b in space_boundaries if b.Name == "2ndLevel"):
    shape = ifcopenshell.geom.create_shape(s, boundary.ConnectionGeometry.SurfaceOnRelatingElement)
    # geom.utils.display_shape(shape)
```
Other tries : 
```python
shape = ifcopenshell.geom.create_shape(s, boundary.ConnectionGeometry)
shape = ifcopenshell.geom.create_shape(s, boundary, boundary.ConnectionGeometry.SurfaceOnRelatingElement)
```
Error message is most of the time : 
```python
  File "./script.py", line 17, in <module>
    shape = geom.create_shape(s, boundary.ConnectionGeometry.SurfaceOnRelatingElement)
  File "/usr/lib/python3.7/site-packages/ifcopenshell/geom/main.py", line 170, in create_shape
    repr.wrapped_data if repr is not None else None
  File "/usr/lib/python3.7/site-packages/ifcopenshell/ifcopenshell_wrapper.py", line 480, in create_shape
    return _ifcopenshell_wrapper.create_shape(settings, instance, representation)
RuntimeError: Failed to process representation item
```
The only success I got was with : 
```python
shape = geom.create_shape(s, boundary.ConnectionGeometry.SurfaceOnRelatingElement.OuterBoundary)
```
But the issue is that it doesn't care about placement / location.

Am I doing something wrong these kind of entity are not yet handled ?

IfcOpenShell version : 0.6.0a1
python : 3.7.3

**Comment by aothms** [comment]
You're correct, this is currently a bit of an oversight. It's not properly implemented because it's not an IfcProduct, but a relationship entity that happens to carry some reference to a geometry.

> But the issue is that it doesn't care about placement / location.

But that's the thing, because this is not a product, it also doesn't have a placement. The documentation mentions something that the geometry is placed within the placement system of the related space (I think), but that relationship is not explicitly there in the schema. What you can do is interpret the geometry of the IfcSpace and reuse that location, because the conversion of a placement is not really exposed in Python.

So, in short, I think you'd be able to get it to work as is, but some more explicit support for space boundary geometry can be added over time.

**Comment by CyrilWaechter** [comment]
Hi,

I followed your advice. It works great when model is exported in meters :
![2020-04-18_11-25](https://user-images.githubusercontent.com/9622862/79634282-d4cd6080-8169-11ea-8a53-43a64aa74df3.png)
What I don't understand is that when I export the model in centimeters (from Revit) I get widely scattered placements : 
![2020-04-18_11-23](https://user-images.githubusercontent.com/9622862/79634341-2249cd80-816a-11ea-9327-be9be5acff32.png)
IfcRelSpaceBoundaries shapes and surfaces seems correct but I can't find the correct way to place it.

The 2 simple test files : [2Storey_2x3_test_files.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/4496419/2Storey_2x3_test_files.zip)

**Comment by aothms** [comment]
> What I don't understand is that when I export the model in centimeters (from Revit) I get widely scattered placements :

My guesses are:
* You implemented support for handling the placements in Python (great, perhaps consider contributing, see #829)
* You didn't take into account the project units there
* You used the default settings in create_shape() which converts everything silently to meters.

So either use `settings.set(settings.CONVERT_BACK_UNITS, True)` to have the geometries correspond to the placement matrices or apply the unit conversion to placement matrices (only IfcCartesianPoint). See below for an example unit lookup (something like that will also be part of the util library (#829 ))

~~~
prefix_mapping = {
    "EXA"  : 1.e18,
    "PETA" : 1.e15,
    "TERA" : 1.e12,
    "GIGA" : 1.e9,
    "MEGA" : 1.e6,
    "KILO" : 1.e3,
    "HECTO": 1.e2,
    "DECA" : 1.,
    "DECI" : 1.e-1,
    "CENTI": 1.e-2,
    "MILLI": 1.e-3,
    "MICRO": 1.e-6,
    "NANO" : 1.e-9,
    "PICO" : 1.e-12,
    "FEMTO": 1.e-15,
    "ATTO" : 1.e-18
}

def get(f, unit_type, default=None):
    us = [u for u in f.by_type("IfcUnitAssignment")[0][0] if u.UnitType == unit_type]
    if len(us) == 0:
        return default
        
    u = us[0]
                
    F = 1.
    if u.is_a("IfcConversionBasedUnit"):
        fc, u = u.ConversionFactor
        F = fc.wrappedValue
        
    assert u.is_a("IfcSIUnit")
    S = prefix_mapping.get(u.Prefix, 1.)

    return F * S

f = ifcopenshell.open(...)
get_unit(f, "LENGTHUNIT")
~~~

**Comment by CyrilWaechter** [comment]
> My guesses are:
> 
>     * You implemented support for handling the placements in Python (great, perhaps consider contributing, see #829)

Unfortunately not. I currently cheat ^^ like following : 
```python
def get_placement(space):
    """Retrieve object placement"""
    space_geom = ifcopenshell.geom.create_shape(BREP_SETTINGS, space)
    # IfcOpenShell matrix values FreeCAD matrix values are transposed
    ios_matrix = space_geom.transformation.matrix.data
    m_l = list()
    for i in range(3):
        line = list(ios_matrix[i::3])
        line[-1] *= SCALE
        m_l.extend(line)
    return FreeCAD.Matrix(*m_l)

space_placement = get_placement(ifc_space)
for ifc_boundary in (b for b in ifc_space.BoundedBy if b.Name == "2ndLevel"):
    try:
        face = RelSpaceBoundary.create(ifc_entity=ifc_boundary)
        second_levels.addObject(face)
        face.Placement = space_placement
```
During RelSpaceBoundary creation I do : 
```python
obj.Shape = create_fc_shape(ifc_entity)
```
function `create_fc_shape` create a shape from `ifc_boundary.ConnectionGeometry.SurfaceOnRelatingElement` if it is created as a brep :
```python
def _part_by_brep(ifc_entity):
    """ Create a Part Shape from brep generated by ifcopenshell from ifc geometry"""
    ifc_shape = ifcopenshell.geom.create_shape(BREP_SETTINGS, ifc_entity)
    fc_shape = Part.Shape()
    fc_shape.importBrepFromString(ifc_shape.brep_data)
    fc_shape.scale(SCALE)
    return fc_shape
```
The `SCALE` factor is there to convert meter to millimeters which is FreeCAD internal unit for length. (`SCALE = 1000`)


> * You didn't take into account the project units there
> 
> * You used the default settings in create_shape() which converts everything silently to meters.

Everything silently converted to meters is exactly what I want but seeing the result I miss something at some point.

**Comment by aothms** [comment]
Matrix math is always confusing to me. But what makes it extra confusing is that the ifcopenshell matrix is a 3x4 matrix (that way only T R S can be encoded, it's the internal structure in OCCT for a gp_Trsf, 3x3 for R S and seperate 1x3 for T).

What about the following code?

```
M = [list(ios_matrix[i*3:i*3+3]) + [1. if i == 3 else 0.] for i in range(4)]
for i in range(3): M[3][i] *= SCALE
flat = sum(M, [])
FreeCAD.Matrix(*flat)
```

If this is not properly transposed, transposition is now 

```
transposed = sum([flat[i::4] for i in range(4)], [])
```

Because I really don't know how FreeCAD is going to handle 12 values.

**Comment by CyrilWaechter** [comment]
> Because I really don't know how FreeCAD is going to handle 12 values.

FreeCAD handle both 3x4 and 4x4 matrix. So this is ok

I very probably found the issue in the only place I manually retrieve position : 
```python
    matrix = get_matrix(ifc_entity.BasisSurface.Position)
    fc_shape = fc_shape.transformGeometry(matrix)
    return fc_shape


def get_matrix(position):
    """Transform position to FreeCAD.Matrix"""
    # FIXME: This doesn't work when unit is not meter
    location = FreeCAD.Vector(position.Location.Coordinates).scale(SCALE, SCALE, SCALE)

    v_1 = FreeCAD.Vector(position.RefDirection.DirectionRatios)
    v_3 = FreeCAD.Vector(position.Axis.DirectionRatios)
    v_2 = v_3.cross(v_1)
```
Apparently `ifc_entity.BasisSurface.Position` is in ifc lenght units (centimeters in test case). So I need to convert it to meters before continuing. I'll take a look at utils to use and contribute.

**Comment by aothms** [comment]
> Apparently ifc_entity.BasisSurface.Position is in ifc lenght units 

Yes, that's likely it. It's a shame this relationship is so implicit in IFC. The direct conversion of express simple types to python data types causes this link to be even further obscured.

**Comment by CyrilWaechter** [comment]
Not a definitive solution as it is not ifcopenshell only but for info an implementation for FreeCAD is available in project I worked on : [BIMxBEM](https://github.com/ENAC-CNPA/BIMxBEM)

**Comment by CyrilWaechter** [comment]
Hi @aothms,
I currently implement a functionality to import IfcRelSpaceBoundary in Blender BIM addon. For my project I used workarounds but it would be great to avoid if it is a bug. I have set up a minimal test case (I probably should convert it into pytest…) first using [Triangle_2x3_R19.ifc](https://github.com/ENAC-CNPA/BIMxBEM/blob/master/IfcTestFiles/Triangle_2x3_R19.ifc) as test file:
```python
import ifcopenshell
import ifcopenshell.geom

settings = ifcopenshell.geom.settings()
ifc_file = ifcopenshell.open("IfcTestFiles/Triangle_2x3_R19.ifc")

for boundary in ifc_file.by_type("IfcRelSpaceBoundary"):
    try:
        ifc_shape = boundary.ConnectionGeometry.SurfaceOnRelatingElement
        geom = ifcopenshell.geom.create_shape(settings, ifc_shape)
        print(f"{boundary}: Success")
    except RuntimeError:
        print(f"{boundary}: Fail")
```
You currently get this output (2/7 success) :
```python
#2999=IfcRelSpaceBoundary('1GVKSPAXv14gcMri0s9vjl',#41,'2ndLevel',$,#168,#344,#197,.PHYSICAL.,.EXTERNAL.): Success
#3001=IfcRelSpaceBoundary('09R2N2LjP0FR0xkSyoPdDs',#41,'2ndLevel',$,#168,#1627,#211,.PHYSICAL.,.INTERNAL.): Fail
#3002=IfcRelSpaceBoundary('1gbc2T7D95owjIQ62vLUpi',#41,'2ndLevel',$,#168,#447,#236,.PHYSICAL.,.EXTERNAL.): Success
#3003=IfcRelSpaceBoundary('3WWI_X3UT8sBwwvwPJt8VH',#41,'2ndLevel',$,#168,#2642,#250,.PHYSICAL.,.EXTERNAL.): Fail
#3004=IfcRelSpaceBoundary('0VJJ1qnvD8lBypx6dAtFGi',#41,'2ndLevel',$,#168,#2723,#266,.PHYSICAL.,.EXTERNAL.): Fail
#3005=IfcRelSpaceBoundary('3e25P8nWn4EhyPjVq9fYLx',#41,'2ndLevel',$,#168,#2811,#278,.PHYSICAL.,.EXTERNAL.): Fail
#3006=IfcRelSpaceBoundary('3ifhrLAVv9cBu73LKAQG8s',#41,'2ndLevel',$,#168,#2776,#292,.PHYSICAL.,.INTERNAL.): Fail
```
Is it a bug ? I can't help that much on the C++ side.

**Comment by aothms** [comment]
`#210= IFCCURVEBOUNDEDPLANE(#209,#204,$);`

I don't know why the error isn't more informative, but InnerBoundaries is not optional. I think that's the issue.

**Comment by CyrilWaechter** [comment]
> InnerBoundaries is not optional

Uh ? [Ifc documentation description](https://standards.buildingsmart.org/IFC/RELEASE/IFC4/ADD2_TC1/HTML/link/ifccurveboundedplane.htm) says it is optional : 
> An optional set of inner boundaries. They shall not intersect each other or the outer boundary.

But I do not know how to interpret STEP specifications:
> InnerBoundaries : SET [0:?] OF IfcCurve;

For optional attributes of a defined entity I see that the `OPTIONAL` prefix is used but is it the same for `SET` ?

**Comment by aothms** [comment]
Myeah they should have said "potentially empty", but it's not optional in
the schema. You could argue that semantically Optional Set [1:?] is roughly
equivalent to NonOptional Set [0:?] they chose the latter.

Sent from a mobile device. Excuse my brevity. Kind regards, Thomas

Op zo 31 jan. 2021 13:33 schreef Cyril Waechter <notifications@github.com>:

**Comment by CyrilWaechter** [comment]
Setting empty InnerBoundaries to `()` fix the issue. I'll report the issue to revit-ifc project which produced this file.

**Comment by CyrilWaechter** [comment]
@aothms I have another one which fails but this time when I use `ifcopenshell.validate` nothing appears wrong.
* file : [failing_boundary.ifc.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/7734157/failing_boundary.ifc.zip)

# Code to reproduce
```python
import ifcopenshell
import ifcopenshell.geom

settings = ifcopenshell.geom.settings()
settings.set(settings.EXCLUDE_SOLIDS_AND_SURFACES, False)
settings.set(settings.USE_BREP_DATA, False)
ifc_file = ifcopenshell.open("failing_boundary.ifc")
boundary = ifc_file.by_id(72)
curve_bounded_plane = boundary.ConnectionGeometry.SurfaceOnRelatingElement
geom = ifcopenshell.geom.create_shape(settings, curve_bounded_plane)
```

# Error message
```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-68dca8a89325> in <module>
----> 1 geom = ifcopenshell.geom.create_shape(settings, curve_bounded_plane)

~/git/IfcOpenShell/src/ifcopenshell-python/ifcopenshell/geom/main.py in create_shape(settings, inst, repr)
    207     return wrap_shape_creation(
    208         settings,
--> 209         ifcopenshell_wrapper.create_shape(settings, inst.wrapped_data, repr.wrapped_data if repr is not None else None),
    210     )
    211 

~/git/IfcOpenShell/src/ifcopenshell-python/ifcopenshell/ifcopenshell_wrapper.py in create_shape(settings, instance, representation)
   1651 def create_shape(settings, instance, representation=None):
   1652     r"""create_shape(IteratorSettings settings, entity_instance instance, entity_instance representation=None) -> boost::variant< IfcGeom::Element *,IfcGeom::Representation::Representation * >"""
-> 1653     return _ifcopenshell_wrapper.create_shape(settings, instance, representation)
   1654 
   1655 def serialise(schema_name, shape_str, advanced=True):

RuntimeError: Failed to process representation item
```

**Comment by Moult** [comment]
For your latest example, it is still invalid. The inner boundaries is coincident with the outer boundaries. The rule for inner boundaries is: "An optional set of inner boundaries. They shall not intersect each other or the outer boundary."

I think this is therefore now resolved.

**Comment by aothms** [comment]
Indeed, we do have the tech now to reliably handle this btw https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcgeom_schema_agnostic/boolean_utils.cpp#L561 Testing for inner-outer bound intersection and falling back to boolean subtraction in that case. But maybe failing is better in this case :)

**Comment by CyrilWaechter** [comment]
> Indeed, we do have the tech now to reliably handle this btw https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcgeom_schema_agnostic/boolean_utils.cpp#L561 Testing for inner-outer bound intersection and falling back to boolean subtraction in that case. But maybe failing is better in this case :)

It would be great to have this fallback automatically or manually as guess which software do it wrong. Currently I had to bake a dirty workaround in each software (BIMxBEM, blenderbim).
